File: src/viz/events.rs
use std::collections::HashMap;
use std::time::Duration;

use mini_aurora_common::{Lsn, PageId};

/// Every discrete internal operation that the visualization can display.
#[derive(Debug, Clone, serde::Serialize)]
pub enum VizEvent {
    // ── PUT path ──────────────────────────────────────────────────────

    /// A new mini-transaction was created.
    MtrCreated { mtr_id: u64, num_records: usize },

    /// LSNs assigned to the records in this batch.
    AssignLsns { first_lsn: Lsn, last_lsn: Lsn },

    /// A record's prev_lsn was linked to the chain for its page.
    LinkPrevLsn { lsn: Lsn, page_id: PageId, prev_lsn: Lsn },

    /// Records appended to the WAL file.
    WalAppend { first_lsn: Lsn, last_lsn: Lsn, offset: u64, bytes: u64 },

    /// WAL fsync completed.
    WalSync,

    /// Page index updated: page_id now points to latest_lsn.
    UpdatePageIndex { page_id: PageId, latest_lsn: Lsn },

    /// LSN→offset mapping recorded.
    UpdateLsnOffset { lsn: Lsn, file_offset: u64 },

    /// VCL advanced.
    AdvanceVcl { old: Lsn, new: Lsn },

    /// VDL advanced (new consistency point).
    AdvanceVdl { old: Lsn, new: Lsn },

    /// Compute read_point updated after a write.
    UpdateReadPoint { old: Lsn, new: Lsn },

    /// Buffer pool entry invalidated after writing to a page.
    BufferPoolInvalidate { page_id: PageId },

    // ── GET path ──────────────────────────────────────────────────────

    /// Buffer pool lookup on the compute side.
    BufferPoolLookup { page_id: PageId, read_point: Lsn, hit: bool },

    /// Storage-side page cache lookup.
    PageCacheLookup { page_id: PageId, read_point: Lsn, hit: bool },

    /// Looked up the page index to find the latest LSN for a page.
    PageIndexLookup { page_id: PageId, latest_lsn: Option<Lsn> },

    /// One step of the prev_lsn chain walk (backwards through WAL).
    ChainWalkStep { page_id: PageId, lsn: Lsn, prev_lsn: Lsn, skipped: bool },

    /// Chain collection complete.
    ChainCollected { page_id: PageId, chain_len: usize, lsns: Vec<Lsn> },

    /// Applying one redo record during materialization.
    MaterializeApply { page_id: PageId, lsn: Lsn, offset: u16, data_len: usize, data_preview: String },

    /// Page materialization complete.
    MaterializeComplete { page_id: PageId, read_point: Lsn },

    /// Materialized page inserted into storage page cache.
    PageCacheInsert { page_id: PageId, read_point: Lsn },

    /// Page inserted into compute buffer pool.
    BufferPoolInsert { page_id: PageId, read_point: Lsn },

    // ── Tiered storage ────────────────────────────────────────────────

    /// A WAL segment was sealed and a new one opened.
    SegmentRotation { sealed_id: u32, new_id: u32, sealed_lsn_range: (u64, u64) },

    /// A read hit a cold-tier segment, incurring extra latency.
    ColdTierRead { segment_id: u32, latency_ms: u64 },

    /// A segment was moved from hot to cold tier.
    SegmentCooled { segment_id: u32 },

    // ── State ─────────────────────────────────────────────────────────

    /// Full system state snapshot for diagram rendering.
    StateSnapshot {
        // Compute state
        node_label: String,
        read_point: Lsn,
        next_mtr: u64,
        buffer_pool_pages: Vec<PageId>,
        // Storage state
        next_lsn: Lsn,
        vcl: Lsn,
        vdl: Lsn,
        page_index: HashMap<PageId, Lsn>,
        lsn_offset_count: usize,
        page_cache_count: u64,
        // WAL state
        wal_file_size: u64,
        wal_lsn_range: Option<(Lsn, Lsn)>,
    },
}

/// Configuration for the visualization system.
#[derive(Debug, Clone)]
pub struct VizConfig {
    /// Delay between each rendered step.
    pub step_delay: Duration,
    /// Whether to use ANSI color codes.
    pub color: bool,
    /// Whether visualization is enabled (can be toggled at runtime).
    pub enabled: bool,
}

impl Default for VizConfig {
    fn default() -> Self {
        Self {
            step_delay: Duration::from_millis(300),
            color: true,
            enabled: true,
        }
    }
}

File: lib/engine/events.ts
// VizEvent -- direct port of src/viz/events.rs + src/viz/metrics.rs

import type { Lsn, PageId } from "./types"

/** Every discrete internal operation that the visualization can display. */
export type VizEvent =
  // PUT path
  | { type: "MtrCreated"; mtrId: number; numRecords: number }
  | { type: "AssignLsns"; firstLsn: Lsn; lastLsn: Lsn }
  | { type: "LinkPrevLsn"; lsn: Lsn; pageId: PageId; prevLsn: Lsn }
  | { type: "WalAppend"; firstLsn: Lsn; lastLsn: Lsn; offset: number; bytes: number }
  | { type: "WalSync" }
  | { type: "UpdatePageIndex"; pageId: PageId; latestLsn: Lsn }
  | { type: "UpdateLsnOffset"; lsn: Lsn; fileOffset: number }
  | { type: "AdvanceVcl"; old: Lsn; new: Lsn }
  | { type: "AdvanceVdl"; old: Lsn; new: Lsn }
  | { type: "UpdateReadPoint"; old: Lsn; new: Lsn }
  | { type: "BufferPoolInvalidate"; pageId: PageId }
  // GET path
  | { type: "BufferPoolLookup"; pageId: PageId; readPoint: Lsn; hit: boolean }
  | { type: "PageCacheLookup"; pageId: PageId; readPoint: Lsn; hit: boolean }
  | { type: "PageIndexLookup"; pageId: PageId; latestLsn: Lsn | null }
  | { type: "ChainWalkStep"; pageId: PageId; lsn: Lsn; prevLsn: Lsn; skipped: boolean }
  | { type: "ChainCollected"; pageId: PageId; chainLen: number; lsns: Lsn[] }
  | {
      type: "MaterializeApply"
      pageId: PageId
      lsn: Lsn
      offset: number
      dataLen: number
      dataPreview: string
    }
  | { type: "MaterializeComplete"; pageId: PageId; readPoint: Lsn }
  | { type: "PageCacheInsert"; pageId: PageId; readPoint: Lsn }
  | { type: "BufferPoolInsert"; pageId: PageId; readPoint: Lsn }
  // State snapshot
  | {
      type: "StateSnapshot"
      nodeLabel: string
      readPoint: Lsn
      nextMtr: number
      bufferPoolPages: PageId[]
      nextLsn: Lsn
      vcl: Lsn
      vdl: Lsn
      pageIndex: Map<PageId, Lsn>
      lsnOffsetCount: number
      pageCacheCount: number
      walByteSize: number
      walLsnRange: [Lsn, Lsn] | null
    }

/** Format an event to a compact one-liner (matches renderer.rs format_one_liner). */
export function formatOneLiner(event: VizEvent): string {
  switch (event.type) {
    case "MtrCreated":
      return `MTR #${event.mtrId} (${event.numRecords} record)`
    case "AssignLsns":
      return event.firstLsn === event.lastLsn
        ? `Assign LSN ${event.firstLsn}`
        : `Assign LSNs ${event.firstLsn}..${event.lastLsn}`
    case "LinkPrevLsn":
      return event.prevLsn === 0
        ? `Link L${event.lsn}(pg${event.pageId}) prev=none`
        : `Link L${event.lsn}(pg${event.pageId}) prev=L${event.prevLsn}`
    case "WalAppend":
      return event.firstLsn === event.lastLsn
        ? `WAL append L${event.firstLsn} @${event.offset} (${event.bytes}B)`
        : `WAL append L${event.firstLsn}..${event.lastLsn} @${event.offset} (${event.bytes}B)`
    case "WalSync":
      return "WAL sync (fsync)"
    case "UpdatePageIndex":
      return `Page index pg${event.pageId}->L${event.latestLsn}`
    case "UpdateLsnOffset":
      return `LSN offset L${event.lsn}->${event.fileOffset}`
    case "AdvanceVcl":
      return `VCL: ${event.old}->${event.new}`
    case "AdvanceVdl":
      return `VDL: ${event.old}->${event.new} (CPL)`
    case "UpdateReadPoint":
      return `read_point: ${event.old}->${event.new}`
    case "BufferPoolInvalidate":
      return `Evict pg${event.pageId} from buffer`
    case "BufferPoolLookup": {
      const tag = event.hit ? "HIT" : "MISS"
      return `BufPool pg${event.pageId} @L${event.readPoint}: ${tag}`
    }
    case "PageCacheLookup": {
      const tag = event.hit ? "HIT" : "MISS"
      return `PageCache pg${event.pageId} @L${event.readPoint}: ${tag}`
    }
    case "PageIndexLookup":
      return event.latestLsn !== null
        ? `PageIdx pg${event.pageId}->L${event.latestLsn}`
        : `PageIdx pg${event.pageId}: not found`
    case "ChainWalkStep":
      return event.skipped ? `Chain: skip L${event.lsn}` : `Chain: collect L${event.lsn}`
    case "ChainCollected": {
      const chain = event.lsns.map((l) => `L${l}`).join("->")
      return `Chain pg${event.pageId}: ${event.chainLen} rec [${chain}]`
    }
    case "MaterializeApply":
      return `Apply L${event.lsn} ${event.dataLen}B @${event.offset}`
    case "MaterializeComplete":
      return `Materialized pg${event.pageId} @L${event.readPoint}`
    case "PageCacheInsert":
      return `Cache pg${event.pageId} @L${event.readPoint}`
    case "BufferPoolInsert":
      return `BufPool insert pg${event.pageId} @L${event.readPoint}`
    case "StateSnapshot":
      return ""
  }
}

/** Aggregate metrics collector -- port of src/viz/metrics.rs. */
export interface MetricsSummary {
  writeCount: number
  readCount: number
  pageCacheHits: number
  pageCacheMisses: number
  bufferPoolHits: number
  bufferPoolMisses: number
  materializeCount: number
  walBytesWritten: number
  vcl: number
  vdl: number
  uptimeSecs: number
}

export class MetricsCollector {
  private writeCount = 0
  private readCount = 0
  private pageCacheHits = 0
  private pageCacheMisses = 0
  private bufferPoolHits = 0
  private bufferPoolMisses = 0
  private materializeCount = 0
  private walBytesWritten = 0
  private vcl = 0
  private vdl = 0
  private startTime = Date.now()

  recordEvent(event: VizEvent): void {
    switch (event.type) {
      case "WalAppend":
        this.writeCount++
        this.walBytesWritten += event.bytes
        break
      case "PageCacheLookup":
        if (event.hit) this.pageCacheHits++
        else this.pageCacheMisses++
        break
      case "BufferPoolLookup":
        if (event.hit) this.bufferPoolHits++
        else {
          this.bufferPoolMisses++
          this.readCount++
        }
        break
      case "MaterializeComplete":
        this.materializeCount++
        break
      case "AdvanceVcl":
        this.vcl = event.new
        break
      case "AdvanceVdl":
        this.vdl = event.new
        break
    }
  }

  summary(): MetricsSummary {
    return {
      writeCount: this.writeCount,
      readCount: this.readCount,
      pageCacheHits: this.pageCacheHits,
      pageCacheMisses: this.pageCacheMisses,
      bufferPoolHits: this.bufferPoolHits,
      bufferPoolMisses: this.bufferPoolMisses,
      materializeCount: this.materializeCount,
      walBytesWritten: this.walBytesWritten,
      vcl: this.vcl,
      vdl: this.vdl,
      uptimeSecs: (Date.now() - this.startTime) / 1000,
    }
  }

  reset(): void {
    this.writeCount = 0
    this.readCount = 0
    this.pageCacheHits = 0
    this.pageCacheMisses = 0
    this.bufferPoolHits = 0
    this.bufferPoolMisses = 0
    this.materializeCount = 0
    this.walBytesWritten = 0
    this.vcl = 0
    this.vdl = 0
    this.startTime = Date.now()
  }
}
